File: Intermediate/26.Solve the Coin Change Problem (minimum coins needed to make a sum)/Solution.py
def coin_change(coins, amount):
    dp = [float('inf')] * (amount + 1)
    dp[0] = 0

    for i in range(1, amount + 1):
        for coin in coins:
            if i >= coin:
                dp[i] = min(dp[i], dp[i - coin] + 1)

    return -1 if dp[amount] == float('inf') else dp[amount]

if __name__ == "__main__":
    n = int(input("Enter the number of coin types: "))
    coins = list(map(int, input("Enter the coin denominations: ").split()))
    amount = int(input("Enter the amount: "))

    result = coin_change(coins, amount)
    if result == -1:
        print("No solution exists.")
    else:
        print("Minimum coins needed:", result)

File: Intermediate/26.Solve the Coin Change Problem (minimum coins needed to make a sum)/Solution.c
#include <stdio.h>
#include <limits.h>

int coinChange(int coins[], int n, int amount) {
    int dp[amount + 1];
    dp[0] = 0;

    for (int i = 1; i <= amount; i++) {
        dp[i] = INT_MAX;
        for (int j = 0; j < n; j++) {
            if (i >= coins[j] && dp[i - coins[j]] != INT_MAX) {
                dp[i] = (dp[i] < dp[i - coins[j]] + 1) ? dp[i] : dp[i - coins[j]] + 1;
            }
        }
    }

    return dp[amount] == INT_MAX ? -1 : dp[amount];
}

int main() {
    int n, amount;
    printf("Enter the number of coin types: ");
    scanf("%d", &n);

    int coins[n];
    printf("Enter the coin denominations:\n");
    for (int i = 0; i < n; i++)
        scanf("%d", &coins[i]);

    printf("Enter the amount: ");
    scanf("%d", &amount);

    int result = coinChange(coins, n, amount);
    if (result == -1)
        printf("No solution exists.\n");
    else
        printf("Minimum coins needed: %d\n", result);

    return 0;
}
